web/assets/js/components/gameoptions.ts

```diff
@@ -1,237 +1,283 @@
+        this.DOM.timerBtns = [...this.DOM.el.querySelectorAll(".js-timerUpdate")];
+        this.DOM.timerDisplay = this.DOM.el.querySelector("#js-timerdisplay");
+        this.DOM.timerIncrement = this.DOM.el.querySelector("#js-timerincrement");
+        this.DOM.ctrl0 = this.DOM.el.querySelector("#js-gameoptions-ctr0");
+        this.DOM.ctrl1 = this.DOM.el.querySelector("#js-gameoptions-ctr1");
+
+        //controls events
+        //TODO: keybord press events
+        this.events.clickOnCtrl0 = this.on({
+            e: "click",
+            target: this.DOM.ctrl0,
+            cb: this._clickOnCtrl0.bind(this),
+        });
+        this.events.clickOnCtrl1 = this.on({
+            e: "click",
+            target: this.DOM.ctrl1,
+            cb: this._clickOnCtrl1.bind(this),
+        });
+
+        //category events
+        this.DOM.categoryBtns.forEach((categoryBtn: Element, i: number) => {
+            this.events["clickOnCategory" + i] = this.on({
+                e: "change",
+                target: categoryBtn,
+                cb: this._updateCategory.bind(this),
+            });
+        });
+
+        //timer events
+        this.DOM.timerBtns.forEach((timerBtn: Element, i: number) => {
+            this.events["clickOnTimer" + i] = this.on({
+                e: "click",
+                target: timerBtn,
+                cb: this._updateTimer.bind(this),
+            });
         });
-      }
+
+        //tl
+        this.switchAnimation1 = gsap
+            .timeline({paused: true})
+            .to(this.DOM.paneConnection, {autoAlpha: 0, display: "none", duration: 0.6})
+            .to(this.DOM.paneCategory, {autoAlpha: 1, display: "flex", duration: 0.6})
+            .to(this.DOM.paneTimer, {autoAlpha: 1, display: "flex", duration: 0.6}, "<");
+
+        this.switchAnimation2 = gsap
+            .timeline({paused: true})
+            .to(this.DOM.paneCategory, {autoAlpha: 0, display: "none", duration: 0.6})
+            .to(this.DOM.paneTimer, {autoAlpha: 0, display: "none", duration: 0.6}, "<")
+            .to(this.DOM.ctrl1, {autoAlpha: 0, duration: 0.6}, "<")
+            .to(this.DOM.paneLoader, {autoAlpha: 1, display: "flex", duration: 0.6});
+
+        //actions
+        this.DOM.timerDisplay.innerHTML = this.options.timer[0];
+        this.DOM.timerIncrement.innerHTML = this.options.timer[1];
+
+        //checkstep
+        Actions.getInstance().then((actions) => {
+            // TODO @Alexis please check that the semantics are right here
+            if (actions.getFaucetToken()) {
+                this._clickOnCtrl1(null, true);
+            }
+
+            this.appear();
+        })
+    }
+
+    _clickOnCtrl0() {
+        this.currentState--;
+
+        if (this.currentState === 1) {
+            this.switchAnimation2.reverse();
+            this.lookingForRival = false;
+            //TODO: stop WS rival finding -> this.lookingForRival
```

Comment:
Maybe add a way to clearTimeout of Action - `waitForGame` method in order to stop rival finder?

Reply:
I honestly have no idea how we can add that kind of support in Typescript, without doing severe gymnastics...

Maybe try something with: `Promise.race`, where the second promise is just a timeout?
```ts
  Promise.race([
    waitForGame(...),
    new Promise((_, reject) =>
      // timeout after x seconds
      setTimeout(() => reject(new Error("timed out")), x)
    ),
  ]);
```

Reply:
I've added support for a timeout, which is probably the cleanest solution:
0e58a7813ba3a16b6bd09bbd7ea12abc6d97a31b